### ISSUE TITLE
[beta] Optional submodules

### ISSUE THREAD

**Issue by mottosso** [mottosso/Qt.py#185]
It's discussion time.

At the moment, in the current latest commit of this repository, lies the 1.0.0 beta of Qt.py. In it, we have implemented #152 via #173 which means Qt.py contains a subset of members from PySide2.

For the past few months, I and others have tested this and it have worked well. Up to a point.

The problem lies in the strict requirement on availability. Meaning, for Qt.py to even import, these submodules must exist in the original binding.

```python
__all__ = [
    "QtGui",
    "QtCore",
    "QtWidgets",
    "QtNetwork",
    "QtXml",
    "QtHelp",
]
```

Which in most cases is fine. They do exist, most of the time. But not always.

`QtHelp` does not exist in Houdini 16 (at the time of this writing), `QtXml` does not exist in the `python-pyqt5` package of Ubuntu 14.10 and god knows what else is missing from other packages.

The result is a fragile Qt.py, that "sometimes works, sometimes doesn't" depending on factors often outside of your control.

### Possible solution

What I've been considering is making some submodules required, and some optional. `QtGui`, `QtCore` and `QtWidgets` are a given, whereas others may be optional.

It would break the current mantra "If it works with Qt.py, it works on any binding" in that it may work on a platform with support for, say, `QtXml` but then break on platforms without it. On the other hand it would be more forgiving on less-often-used submodules, such as `QtHelp`.

Another option is to give up on strict enforcement, and keep things are they are. The most forgiving, but also least admitting to potential threats of cross-binding compatibility.

**Comment by tomwod** [comment]
hey.  imo i would prefer optional over strict if it a binary choice.  
how about by default be strict, as is, which is explicit.  but allow optional submodules via environment setting - so I can explicitly acknowledge that something is missing in a context that I'm responsible for?

**Comment by mottosso** [comment]
Thanks @tomwod, that's a good suggestion.

**Comment by davvid** [comment]
I would very much like these sometimes-available module to be exposed by Qt.py.  It would be better to not need any environment variables or configuration whatsoever.

It's a good idea to push the burden of dealing with sometimes-available modules to the app developer rather than the global environment.  Apps can then selectively use features[1] and handle the ImportError app-side when importing the sometimes-available module from Qt.py fails.

I would suggest taking it all the way and allowing for QtWebEngineWidgets and QtOpenGL to conditionally exist in the Qt.py module.  There's really no benefit in being strict besides sticking with the lowest common denominator, which makes it less useful and hurts adoption.

Here's an example where cola does exactly this using qtpy:
[1] https://github.com/git-cola/git-cola/blob/master/cola/widgets/about.py#L377

IMO this is better than an environment variable because that's not really a practical solution. Setting the variable won't magically fix the app code, so it makes more sense for the app to make that decision. Making it work in such environments requires the app to handle the sometimes-available functionality either way, so having the environment variable just adds another dimension of complexity / things-to-test.

It'd be better to not have an option at all, and just make Qt.py strictly follow the capabilities of the underlying binding.

Here is what Qt.py would require apps to do to handle a sometimes-available webkit module:

```python
import Qt

if Qt.__binding__ == 'PyQt4':
    from PyQt4.QtWebKit import QWebView as QWebEngineView
elif Qt.__binding__ == 'PyQt5':
    from PyQt5.QtWebKitWidgets import QWebEngineView
elif Qt.__binding__ == 'PySide':
    from PySide.QtWebKit import QWebView as QWebEngineView
elif Qt.__binding__ == 'PySide2':
    QWebEngineView = None  # means it's missing
```

Compare this to the equivalent `qtpy` code:

```python

import qtpy
try:
    from qtpy.QtWebEngineWidgets import QWebEngineView
except (ImportError, qtpy.PythonQtError):
    QWebEngineView = None  # means it's missing
```

It would be really nice for Qt.py to follow suite and allow apps to simplify their code to this:
```
import Qt
try:
    from Qt.QtWebEngineWidgets import QWebEngineView
except ImportError:
    QWebEngineView = None  # means it's missing
```

One way to still allow for Qt.py to protect users with its strict subset would be to make that strictness opt-in.  Setting something like `QT_PY_LEAN_AND_MEAN=1`  would make Qt.py avoid all sometimes-available modules, and would then allow apps to test stuff out and selectively handle the `ImportErrors`.

This seems like a good pragmatic option because it does not hurt the common case while still empowering app developers to suss out the binding-specific differences while they develop by setting an environment variable.

Making the sometimes-available modules opt-in is actually a big burden for users.  Now, not only do app developers need to change their code (which is a given), but the end user also has to set an environment variable.  That's a burden for developers too, because now we have to document it and tell users about it, whereas not having an option at all is just plain simpler.

Another reason to make the strict `QT_PY_LEAN_AND_MEAN` mode opt-in is that it's sometimes hard to know exactly what has been disabled in different toolkits, so Qt.py's definition of "common subset" may not necessarily reflect reality, so it's safer for the code to safely handle cases where stuff is missing, even if it's unexpected, so that unforeseen omissions in binding symbols are robustly handled.

Thanks for listening.

**Comment by mottosso** [comment]
Thanks @davvid!

The `QT_PY_LEAN_AND_MEAN` environment variable approach was actually the first iteration of this, except we called it `QT_STRICT`.

- See [here](https://github.com/mottosso/Qt.py/blob/1ad7ecc1740f411fefae24c8a8f92082e49ab57a/Qt.py#L927).

To your example, how about this?

```python
import Qt

if Qt.__binding__ in ('PyQt4', 'PyQt5', 'PySide') :
    from Qt.QtWebKit import QWebView as QWebEngineView
else:
    QWebEngineView = None  # means it's missing
```

I agree the git-cola example does look a little neater and also handles cases where a binding may exist, but on some platform lack a particular submodule or member of said submodule.

<br>

It's a tough choice, there's tension between (1) exposing members that may not exist everywhere in order make it less frightening for new users and (2) offering a strict subset that the seasoned developer can expect to bark loudly about incompatibilities.

Personally I want my software to work everywhere and constantly bend over backwards to make it happen. Most of the time, I won't know when to do to that until after it has hit the end-user on some obscure version of a binding and this strictness is what would enable me to see this up-front all the time.

I think best we can do is get all of the pros and cons out in the open so we can make an educated decision. I also think there's room for more than one wrapper to Qt, room for QtPy, and that we can let our decisions differ from theirs so the end user can make a choice based on what works best for them.

**Comment by davvid** [comment]
Thanks @mottosso your example looks like a very nice way to handle it.

**Comment by fredrikaverpil** [comment]
It's a bit discomforting that e.g. Houdini is missing its `Qt.QtHelp` and therefore will make Qt.py 1.0.0b raise an ImportError. Or, umm, wait, does it just raise a warning? (I'm a Maya knight still, so I wouldn't know)

I believe the whole strictness idea on my end was primarily around certain members and perhaps not so much about entire submodules.

How about if Qt.py would warn the user about an expected submodule (such as `Qt.QtHelp`) not existing. But without raising an error and exiting.

Extra work would then be required if a certain member in a specific binding was missing. And by extra work I mean the "if this binding than that code" approach. I really like this approach as it becomes very clear in the source code what's going on.

I'd like to avoid the usage of stuff like `QT_PY_LEAN_AND_MEAN` (`QT_STRICT`).

**Comment by mottosso** [comment]
> Or, umm, wait, does it just raise a warning?

At the moment, it doesn't raise anything and just excludes Qt.QtHelp when `import Qt`. Trying to import it explicitly - `import Qt.QtHelp` - would result in `ImportError`.

> But without raising an error and exiting.

That wouldn't work, as the application would in that case be the one to throw the exception, and as @davvid pointed out, the application developer will need some way of knowing this so he can handle said exception.

```python
# my_app.py
from Qt import QtHelp

# Expecting the above to not fail, a warning wouldn't prevent this from failing.
QtHelp.doThings()
```

> Extra work would then be required if a certain member in a specific binding was missing. 

This is [already done](https://github.com/mottosso/Qt.py/pull/186/files#diff-cd1d7c0a982bce680146d5b248151256R980) in the current PR #186, i.e. 1.0.0.b. Individual members are simply excluded and would throw an `ImportError` is one attempted to import it explicitly.

Give the current PR as try and let me know what you think.

**Comment by mottosso** [comment]
> At the moment, it doesn't raise anything and just excludes Qt.QtHelp when import Qt. Trying to import it explicitly - import Qt.QtHelp - would result in ImportError.

Having said that, it *does* throw an `ImportError` in the current [master](https://github.com/mottosso/Qt.py/tree/595a3ca514ac7be992a74057306b73603f436996), as it expected `QtHelp` to exists before even allowing you to `import Qt`.

**Comment by fredrikaverpil** [comment]
> At the moment, it doesn't raise anything and just excludes Qt.QtHelp when import Qt.

I think that's good behavior. Are you doing anything so that you could detect a missing module on `import Qt` and possibly raise a warning -- and do you think it would be useful?

> That wouldn't work, as the application would in that case be the one to throw the exception, and as @davvid pointed out, the application developer will need some way of knowing this so he can handle said exception.

Right, right. That makes perfect sense.

> Give the current PR as try and let me know what you think.

Doing it now :)

**Comment by mottosso** [comment]
> I think that's good behavior. Are you doing anything so that you could detect a missing module on import Qt and possibly raise a warning -- and if so, do you think it would be useful?

Not missing, but present.

```python
import Qt

if "QtHelp" in Qt.__all__:
  from Qt import QtHelp

# Which is the same as
try:
  from Qt import QtHelp
except ImportError:
  pass
```

For members..

```python
from Qt import QtWidgets

if "QApplication" in QtWidgets.__dict__:
  app = QtWidgets.QApplication([])
```

Is that the best approach? Not sure. Do we need it? Yes, I do believe so.

Do we need the ability to detect **missing** members, as in..

```python
if "QApplication" in QtWidgets.__missing__:
  raise Exception("Sorry, my program requires QApplication")
```

..I'm not too sure. I can't see the use for it right off the bat, but would be open to suggestions.

**Comment by fredrikaverpil** [comment]
By missing, I meant for instance the Qt.Help in Houdini which is expected (as it is part of PySide2) but is __missing__ in Houdini.

If we were able to produce a warning about this, this would give the developer an early heads up about something which might (or might not) be a deal breaker.

However, it's a problem that Qt.py could never fix either way you look at it... so I'm not sure if it's even useful to produce such warnings.

**Comment by mottosso** [comment]
> I would suggest taking it all the way and allowing for QtWebEngineWidgets and QtOpenGL to conditionally exist in the Qt.py module. There's really no benefit in being strict besides sticking with the lowest common denominator, which makes it less useful and hurts adoption.

I've been thinking about this, and I think what we have got at the moment is exactly that.

```python
try:
    from Qt import QtHelp
except (ImportError):
    # don't punish the users
    webbrowser.open_new_tab('file://' + hotkeys_html)
    return
```

As in, `QtHelp` would be available, if the original binding didn't grant access to it. In the git-cola example, it's about `QtWebEngineWidgets` which isn't and can't be available since it's only available in Qt 5.

**Comment by mottosso** [comment]
> it's about QtWebEngineWidgets which isn't and can't be available since it's only available in Qt 5.

A thought occurred: would it be possible to expose all members of both Qt 4 *and* 5 in this way? Including `QtQml` and other non-Qt4 libraries? Could we use Qt.py for building Qt 5 applications to?

**Comment by davvid** [comment]
I personally like the forward-looking approach taken by qtpy (which is why it has a `QtWebEngineWidgets` module).  It makes the older PyQt4 bindings look like PyQt5, which is nice because we then just write our code assuming a PyQt5-like setup.  The philosophy there is that we want the abstraction layer reflect where we'd like to get (PyQt5) while being able to do so in a PyQt4 world.

**Comment by fredrikaverpil** [comment]
For me personally, I don't want to create new code today which will trap me inside of pre-Qt5 behavior for newly written code.

Instead, what I'm looking for in Qt.py is to code new projects with Qt5 in mind and easily provide backwards compatibility to Qt4.

**Comment by mottosso** [comment]
> The philosophy there is that we want the abstraction layer to reflect where we'd like to get (PyQt5) while being able to do so in a PyQt4 world.

> what I'm looking for in Qt.py is to code new projects with Qt5 in mind and easily provide backwards compatibility to Qt4.

To clarify, this is what we have currently with Qt.py, including with this issue implemented? If you could use `QtWebEngineWidgets`, you would eliminate Qt 4 compatibility.

**Comment by fredrikaverpil** [comment]
I guess what it comes down to is this:
* Are you porting your Qt4 code into running in Qt5 or are you writing new code in Qt5 which you'd like to run in Qt4 also?
* Which of these two should be kept in mind for Qt.py development?

**Comment by fredrikaverpil** [comment]
> To clarify, this is what we have currently with Qt.py, including with this issue implemented?

Yes. I believe so too.

**Comment by mottosso** [comment]
> writing new code in Qt5 which you'd like to run in Qt4 also?

This.

Though there is a new question we need to ask ourselves at this point. If `QtWebEngineWidget` and other Qt 5 friends are exposed, we aren't as concerned with cross-compatibility, as we are with using Qt.py for all bindings.

That's a subtle difference. If we expose `QtWebEngineWidget`, we enable using Qt.py for projects only involving Qt 5. In those cases, maybe we aren't interested in cross-compatibility, we are merely interested in sticking to one Qt library in all of our code.

It'd be a different philosophy, maybe a right one, maybe not.

At the moment, all of your Qt.py code will run identically on all bindings. With qtpy, as it exposes `QtWebEngineWidget`, that isn't the case. Maybe this should be what separates the two? Cross-compatibility versus wrapper.

**Comment by fredrikaverpil** [comment]
> If we expose `QtWebEngineWidget`, we enable using Qt.py for projects only involving Qt 5. In those cases, maybe we aren't interested in cross-compatibility, we are merely interested in sticking to one Qt library in all of our code.

I think that these are two use cases which we should try to have Qy.py support.

### Example 1
I'm maintaining a large application which needs to run in both Maya, Nuke and as well as a standalone application. Maya dictates that I need both Qt4 and Qt5 support, so Qt.py is the preferred solution here.

### Example 2
I'm also developing an application which will only use Qt5. However, it's using PySide2 inside of Maya and PyQt5 when run outside of Maya. In this case, it'll also be helpful to use Qt.py, if possible.

If we could make Qt.py cater to both these scenarios, I would be super happy.

**Comment by mottosso** [comment]
That is possible, but we lose one thing: The alerts currently given when using something unsupported by any binding.

At the moment, if you try and write software to work with Qt 4 and Qt 5, and you use `QtWebEngineWidget`, Qt.py will say "Stop!" and you will know that you cannot use that if your program is to run on all bindings.

If we include `QtWebEngineWidget`, we lose those alerts, but gain the ability to write for Qt 5-only software.

**Comment by fredrikaverpil** [comment]
Thanks for this clarification. Helps me better understand how it works now since I've been away from the Qt.py code for some time.

I know we've been down this road before, but could the developer somehow dictate which "mode" he/she would want to run Qt.py in?
Meaning; set a parameter to receive the stop messages OR support Qt5 development.

Or do you feel we need to settle on one of these behaviors?

**Comment by mottosso** [comment]
> Or do you feel we need to settle on one of these behaviors?

I'm not sure. We might not be able to be everything to everyone.

> I know we've been down this road before, but could the developer somehow dictate which "mode" he/she would want to run Qt.py in?

The trouble is where to specify this mode.

What if a user runs Maya 2016 with two applications; one with the parameter set, and the other without. We couldn't use a global environment variable for that. And we can't set anything on Qt.py before first importing it.

**Comment by fredrikaverpil** [comment]
Yeah, sounds like a mess.

**Comment by mottosso** [comment]
Released as `1.0.0.b2`. (beta)

```bash
$ pip install Qt.py==1.0.0.b2
```
